agent/eni/manager.go

```diff
@@ -0,0 +1,313 @@
+// Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package eni
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"github.com/vishvananda/netlink"
+
+	"github.com/aws/amazon-ecs-agent/agent/eni/netlinkWrapper"
+	log "github.com/cihub/seelog"
+)
+
+const (
+	ethPrefix                     = "eth"
+	defaultReconciliationInterval = time.Second * 30
+	invalidDeviceMsg              = "Invalid Device Name"
+	invalidMACMsg                 = "Invalid MAC Address"
+)
+
+// NOTE: This eases testing with mock
+var sysfsNetDir = "/sys/class/net"
+
+//type enis map[string]string
+
+// Manager exposes the methods to initialize and update ENI's
+// attached to the instance.
+type Manager interface {
+	InitStateManager() error
+	BeginENIUpdate(ctx context.Context)
+	GetAllENIs() map[string]string
+}
+
+// StateManager maintains the state of ENI's connected
+// to the instance. It also has supporting elements to
+// maintain consistency and update intervals
+type StateManager struct {
+	updateLock           sync.RWMutex
+	updateIntervalTicker *time.Ticker
+	enis                 map[string]string // MAC => Device-Name
+	watcher              *fsnotify.Watcher
+	netlinkClient        netlinkWrapper.NetLink
+}
+
+// NewENIManager instanciates a new ENIStateManager
+func NewENIManager() Manager {
+	return newStateManager()
+}
+
+func newStateManager() *StateManager {
+	return &StateManager{
+		enis:          make(map[string]string, 10),
+		netlinkClient: netlinkWrapper.NetLinkClient{},
+	}
+}
+
+// InitStateManager initializes a new ENI State Manager
+func (eniStateManager *StateManager) InitStateManager() error {
+	links, err := eniStateManager.netlinkClient.LinkList()
+	if err != nil {
+		log.Errorf("Error retrieving network interfaces: %s", err.Error())
+		return err
+	}
+
+	for _, link := range links {
+		deviceName, MACAddress := link.Attrs().Name, link.Attrs().HardwareAddr.String()
+		if strings.HasPrefix(deviceName, ethPrefix) {
+			// FIXME: Check if lock necessary ?
```

Comment:
I think we need a lock at the beginning of this method. In case others want to change this map?